src/shared/state/localstate.h

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef ADDONLOCALSTATE_H
+#define ADDONLOCALSTATE_H
+
+#include "settingsholder.h"
+#include "statebase.h"
+
+/**
+ * @brief Local State implementation,
+ * in which state does persist throughout restarts.
+ *
+ * The storage backend for the LocalState is the SettingsHolder.
+ *
+ * The QSettings object in SettingsHolder can be directly accessed by this
+ * class, because it is a friend class of the SettingsHolder class.
+ */
+class LocalState final : public StateBase {
+ public:
+  ~LocalState();
+
+  /**
+   * @brief Construct an SessionState object from a JSON object.
+   *
+   * @param storageIdentifier The key under which this state will be stored in
+   * the local storage backend. This should be unique. A warning will be logged
+   * when another LocalState has already been created with the same identifier
+   * during a given application run.
+   * @param initialState  A JSON object specifying the state schema.
+   * @return LocalState the generated state.
+   */
+  LocalState(const QString& storageIdentifier, const QJsonObject& initialState);
+
+ private:
+  QJsonValue getInternal(const QString& key) const override;
+  void setInternal(const QString& key, const QJsonValue& value) override;
+  void clearInternal(const QString& key = "") override;
+
+  /**
+   * @brief Gets the settings from the settings holder. This class is a friend
+   * of that class, so it can access the internal QSettings directly.
+   *
+   * @return QSettings The application's internal settings.
+   */
+  static QSettings* settings() {
```

Comment:
Why do we need this? Can we remove it?
If by mistake we make it public, we expose the underlying QSettings obj to everybody.

Reply:
Ah, we need it :/ We need to grab the QSettings from the header file. The cpp file is not in the same scope and the `friend` thing doesn't work. Do you have a better suggestion?